fix(aio): make search results better

petebacondarwin opened the pull request:
* update to latest version of lunr search
* add trailing wildcard to search terms to increase matches
* fix unwanted error when escape was pressed

Closes #17417

petebacondarwin commented on aio/src/app/app.component.html:
```diff
@@ -11,7 +11,7 @@
   <a class="nav-link home" href="/"><img src="{{ homeImageUrl }}" title="Home" alt="Home"></a>
   <aio-top-menu *ngIf="isSideBySide" [nodes]="topMenuNodes"></aio-top-menu>
-  <aio-search-box class="search-container" #searchBox (search)="doSearch($event)"></aio-search-box>
+  <aio-search-box class="search-container" #searchBox (onSearch)="doSearch($event)"></aio-search-box>
```
It turned out that because the input box inside the search box component was of type "search" then it would also trigger this handler directly, which meant that the `$event` was an event object rather than the expected string.
The easiest solution was just not to use the name `search` for the component's output event.

petebacondarwin commented on aio/src/app/search/search-worker.js:
```diff
@@ -6,21 +6,22 @@
     this.field('titleWords', {boost: 50});
     this.field('members', {boost: 40});
     this.field('keywords', {boost: 20});
+    addFn(this);
```
The new version of Lunr requires the index to be immutable, so you must load up the index data when you create the index.

petebacondarwin commented:
TODO: I think we should also do the following (after we go to production):

1) Host our own copy of the lunr library, rather than relying upon the unpkg CDN.
2) Prebuild the index on the server, as part of the dgeni build

mary-poppins commented:
The angular.io preview for 0cdac5352993439b393e1909f5b83e1a5552f6e4 is available [here][1].

[1]: https://pr17419-0cdac5352993439b393e1909f5b83e1a5552f6e4.ngbuilds.io/

gkalpak commented on aio/src/app/search/search-worker.js:
```diff
@@ -67,16 +68,26 @@ function makeRequest(url, callback) {
 
 // Query the index and return the processed results
 function queryIndex(query) {
+  var terms = query.split(/\s/);
```
Nit: `/\s+/`

gkalpak reviewed (changes requested):
You also need to update [ngsw-manifest.json][1].

[1]: https://github.com/petebacondarwin/angular/blob/0cdac5352993439b393e1909f5b83e1a5552f6e4/aio/ngsw-manifest.json#L9

MartinKolarik commented on aio/src/app/search/search-worker.js:
```diff
@@ -6,21 +6,22 @@
 
 // NOTE: This needs to be kept in sync with `ngsw-manifest.json`.
-importScripts('https://unpkg.com/lunr@0.7.2/lunr.min.js');
+importScripts('https://unpkg.com/lunr@2.1.0/lunr.js');
```
jsDelivr with on-demand minification to the rescue 😄
https://cdn.jsdelivr.net/npm/lunr@2.1.0/lunr.min.js

petebacondarwin replied:
Don't know. but this is another good reason to host our own?

IgorMinar replied:
yeah. we should host our own

gkalpak commented on aio/src/app/search/search-worker.js:
```diff
@@ -67,16 +68,26 @@ function makeRequest(url, callback) {
+    terms.forEach(function(term) {
+      // Only include terms that are longer than 2 characters, if there is more than one term
+      // Add trailing wildcard to each term so that it will match more results
+      if (terms.length === 1 || term.length > 2) qb.term(term, { wildcard: lunr.Query.wildcard.TRAILING });
```
By doing the `.length > 2` check here, we end up showing the "No results" message when the user starts typing, which is confusing. Could we do the check on the search-box component and not send a query if the term doesn't meet some criteria (e.g. min length)?

gkalpak replied:
I also noticed that we are sending several empty-string `query`. Probably a good idea to avoid or short-circuiting those.

petebacondarwin replied:
But the `terms === 1` should cause the search to go ahead for the first two chars. I suspect that you are just seeing the "No results" because the index has not yet finished building.

gkalpak replied:
Ah, true about the `terms === 1` check. It's probably the index (in which case we should show a better message (but not something for this PR)).

gkalpak commented on aio/src/app/search/search-box/search-box.component.ts:
```diff
@@ -17,15 +17,14 @@ import { LocationService } from 'app/shared/location.service';
-    (keyup)="onSearch($event.target.value)"
-    (focus)="onSearch($event.target.value)"
-    (click)="onSearch($event.target.value)">`
+    (keyup)="doSearch()"
```
I just remembered `(input)` is useful for catching changes coming from right-click+cut/paste actions.

petebacondarwin replied:
Hmm, OK. So I think I need to make all this a bit more observable :-)

gkalpak commented:
> Host our own copy of the lunr library, rather than relying upon the unpkg CDN.

Adding some context:
We have briefly considered that in the past. The problem was that the cli didn't support copying something outside the `src/` directory (i.e. from `aio/node_modules/lunr/`), so we decided to keep fetching drom unkpg.com (since the ServiceWorker will serve it after the first time anyway), until we have time to find a better solution. Definitely something to look into post-release 👍 

##
> Prebuild the index on the server, as part of the dgeni build

I thought the reason we didn't do that was that the index was pretty large for downloading, so it was preferrable to download the keywords and build the index on the client. Has that changed with lunr 2.x?

petebacondarwin commented:
Apparently the index is smaller in 2.x.

mary-poppins commented:
The angular.io preview for bdd895c5565bb63a8c6f1906ff8dba8e845e7336 is available [here][1].

[1]: https://pr17419-bdd895c5565bb63a8c6f1906ff8dba8e845e7336.ngbuilds.io/

IgorMinar reviewed (changes requested):
Let's figure out how to get rid of the dependency on the external cdn.

Unless I'm missing something, I don't see this change as a critical improvement. If that's the case then let's defer this upgrade until later. We should not be introducing this kind of change now. Especially since it's unclear how the non-minified download affects performance.

petebacondarwin committed:
fix(aio): use locally hosted lunr library

The library is downloaded from npm but then
copied into the assets folder (and ignored by git)
as part of the postinstall step.

mary-poppins commented:
The angular.io preview for 8202bd924ea0afae9011e21fa51aa0134a3bfc76 is available [here][1].

[1]: https://pr17419-8202bd924ea0afae9011e21fa51aa0134a3bfc76.ngbuilds.io/

petebacondarwin commented:
I have arranged for us to minify and host the lunr library file ourselves. See the second commit.

I looked into building the index on the server but it is coming in at around 1.5Mb, which is simply too large. So we will continue to build in a web-worker.

Later this week I can look at delaying the download of the keywords, etc till the app is up and running.

petebacondarwin commented on aio/ngsw-manifest.json:
```diff
@@ -5,8 +5,7 @@
       {"url": "https://fonts.gstatic.com/s/materialicons/v22/2fcrYFNaTjcS6g4U3t-Y5ZjZjT5FdEJ140U2DJYC3mY.woff2"},
-      {"url": "https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css"},
-      {"url": "https://unpkg.com/lunr@0.7.2/lunr.min.js"}
+      {"url": "https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css"}
```
I believe we don't need a line for the lunr library now that it is a local file - similar to the prettify.js library, right @gkalpak ?

gkalpak replied:
Right!

petebacondarwin commented on aio/src/app/search/search-box/search-box.component.ts:
```diff
@@ -17,15 +20,19 @@ import { LocationService } from 'app/shared/location.service';
-  @Output() search = new EventEmitter<string>();
+  @Output() onSearch = this.searchSubject
+                           .filter(value => !!(value && value.trim()))
+                           .distinctUntilChanged();
```
Now we only trigger an `onSearch` output event when the query is not blank and is different from before.

petebacondarwin committed:
fix(aio): show search results when search box gets focus

Due to a previous commit, the search was only triggered
if the query changed, and not when the search box regained
focus.

mary-poppins commented:
The angular.io preview for 07483ddd36c12dcb062f0c278ca253ac7191d899 is available [here][1].

[1]: https://pr17419-07483ddd36c12dcb062f0c278ca253ac7191d899.ngbuilds.io/

gkalpak commented on aio/src/app/app.component.spec.ts:
```diff
@@ -723,6 +723,13 @@ describe('AppComponent', () => {
           expect(component.showSearchResults).toBe(false);
         });
+
+        it('should re-run the search when the search box regains focus and is not empty', () => {
```
How is the "is not empty" checked?

gkalpak commented on aio/src/app/search/search-box/search-box.component.ts:
```diff
@@ -17,15 +20,20 @@ import { LocationService } from 'app/shared/location.service';
-  @Output() search = new EventEmitter<string>();
+  @Output() onSearch = this.searchSubject
+                           .distinctUntilChanged()
+                           .filter(value => !!(value && value.trim()));
```
One side effect of this is that it will not clear the search results when you empty the search field (e.g. via BACKSPACE or by clicking on the `x` button). I liked that "feature". Now, the only way to hide the results is to click somewhere outside of the results panel and there is not much space left; espacially on mobile.

petebacondarwin replied:
OK, so let's ditch the filter, then.

petebacondarwin replied:
Although hitting escape does clear the results

gkalpak replied:
Most mobile keyboards don't have escape though :grin:

gkalpak reviewed (commented):
A couple of minor comments.
LGTM otherwise.

mary-poppins commented:
The angular.io preview for 624acdc3d72f605e7f629c30a4f57c81003c69fe is available [here][1].

[1]: https://pr17419-624acdc3d72f605e7f629c30a4f57c81003c69fe.ngbuilds.io/

petebacondarwin commented:
I fixed up the removal of the results view when you empty the search box @gkalpak PTAL

petebacondarwin committed:
fix(aio): make search results better

* update to latest version of lunr search
* add trailing wildcard to search terms to increase matches
* fix unwanted error when escape was pressed

Closes #17417

mary-poppins commented:
The angular.io preview for fec11b141a55766bbf38af84c09d3c16d95c410f is available [here][1].

[1]: https://pr17419-fec11b141a55766bbf38af84c09d3c16d95c410f.ngbuilds.io/

petebacondarwin commented:
Travis failure is in core e2e tests and unrelated to this change

olivernn commented:
> I looked into building the index on the server but it is coming in at around 1.5Mb, which is simply too large. So we will continue to build in a web-worker.

Getting the index size down is my current performance focus. Out of interest, how small would the index need to be for pre-building the index to be a viable solution here?

Also, what is the best way for me to get my hands on your corpus, so I can do some testing locally?

petebacondarwin commented:
@olivernn - thanks for your continued support with this. I really appreciate your proactive approach.
The corpus is a little tricky to acquire because it is generated on the fly as part of our build process for https://github.com/angular/angular/tree/master/aio

One thing you can do is get hold of the data we are using on the client to generate the index, which is https://angular.io/generated/docs/app/search-data.json

And you can see what we do with it in our web worker: https://github.com/angular/angular/blob/master/aio/src/app/search/search-worker.js

If you want to try building the website locally you need to do something along the following lines:

```bash
git clone https://github.com/angular/angular
cd angular
npm install
./build.sh #possibly might not work (on Windows) and may not be necessary
cd aio
yarn setup
yarn start
```
